File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.cpp
#include <queue>
#include <map>
#include <unordered_set>
#include <string>

class Solution {
public:
    int minMutation(string start, string end, vector<string>& bank) {
        
        map<string, unordered_set<string>> m;
        queue<string> q;
        int s = 0;
        m[start] = unordered_set<string>();
        for(string a : bank)
        {
            m[a] = unordered_set<string>();
            if(oneOff(a, start))
            {
                m[start].insert(a);
                m[a].insert(start);
            }
            for(string b: bank)
            {
                if(oneOff(a, b))
                {
                    m[b].insert(a);
                    m[a].insert(b);
                }
            }
        }
        q.push(start);
        int r = 0;
        unordered_set<string> v;
        while(!q.empty())
        {
            int sz = q.size();
            for(int i = 0; i < sz; i++)
            {
                string n = q.front();
                q.pop();
                
                if(n == end)
                    return r;
                    
                for(string a : m[n])
                    if(!v.count(a))
                        q.push(a);
                
                v.insert(n);
            }
            r++;
        }
        return -1;
    }
    
    bool oneOff(string a, string b) {
        bool x = false;
        for(int i = 0; i < 8; i++)
        {
            if(a[i] != b[i])
            {
                if(x)
                    return false;
                x = true;
            }
        }
        return x;
    }
};

File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.cs
public class Solution {
    public int MinMutation(string start, string end, string[] bank) {
        Dictionary<string, HashSet<string>> m = new Dictionary<string, HashSet<string>>();
        Queue<string> q = new Queue<string>();
        m[start] = new HashSet<string>();
        foreach(string a in bank)
        {
            if(!(m.ContainsKey(a)))
                m[a] = new HashSet<string>();
            
            if(oneOff(start, a))
            {
                m[start].Add(a);
                m[a].Add(start);
            }
            foreach(string b in bank)
            {
                if(!(m.ContainsKey(b)))
                    m[b] = new HashSet<string>();
                
                if(oneOff(a, b))
                {
                    m[b].Add(a);
                    m[a].Add(b);
                }
            }
        }
        
        HashSet<string> v = new HashSet<string>();
        
        int r = 0;
        
        q.Enqueue(start);
        while(q.Count() != 0)
        {
            int sz = q.Count();
            for(int i = 0; i < sz; i++)
            {
                string a = q.Dequeue();
                
                if(a == end)
                    return r;
                
                v.Add(a);
                
                foreach(string x in m[a])
                    if(!v.Contains(x))
                        q.Enqueue(x);
            }
            r++;
        }
        return -1;
    }
    
    private bool oneOff(string a, string b)
    {
        bool x = false;
        for(int i = 0; i < 8; i++)
        {
            if(a[i] != b[i])
                if(x)
                    return false;
                else
                    x = true;
        }
        return x;
    }
}